packages/dataviews/src/dataviews-layouts/utils/useInfiniteScrollData.ts

```diff
@@ -0,0 +1,277 @@
+
+						if ( view.infiniteScrollEnabled ) {
+							// Check if this record already has a position
+							const existingPosition =
+								positionMapRef.current.get( itemId );
+							if ( existingPosition !== undefined ) {
+								position = existingPosition;
+							} else {
+								// Assign new position and increment for next record
+								position = nextPosition;
+								positionMapRef.current.set( itemId, position );
+								nextPosition++;
+							}
+						}
+
+						return {
+							...record,
+							position,
+						};
+					} );
+
+				if ( newRecords.length === 0 ) {
+					return prev;
+				}
+
+				// Update the loaded range
+				const allRecords =
+					scrollDirection === 'up'
+						? [ ...newRecords, ...prev ]
+						: [ ...prev, ...newRecords ];
+
+				const allIds = allRecords
+					.filter( ( r ): r is Item => r !== null )
+					.map( ( r ) => r.id );
+				const newMin = Math.min( ...allIds );
+				const newMax = Math.max( ...allIds );
+
+				loadedRangeRef.current = {
+					min: newMin,
+					max: newMax,
+				};
+
+				// Create array with placeholders to maintain positions
+				const result: ( Item | null )[] = [];
+				for ( let id = newMin; id <= newMax; id++ ) {
+					const record = allRecords.find(
+						( r ) => r !== null && r.id === id
+					);
+					result.push( record || null );
+				}
+
+				// Filter to keep only records that should remain visible
+				// Keep items within a certain range of visible entries
+				if ( visibleEntries.length > 0 ) {
+					const visibleMin = Math.min( ...visibleEntries );
+					const visibleMax = Math.max( ...visibleEntries );
+					const buffer = 6;
+
+					const filtered = result
+						.map( ( record, index ) => {
+							const itemId = newMin + index;
+							// Keep records that are null (placeholders) or within the visible range
+							if ( record === null ) {
+								return record;
+							}
+							// Keep items within buffer range of visible items
+							if (
+								itemId >= visibleMin - buffer &&
+								itemId <= visibleMax + buffer
+							) {
+								return record;
+							}
+							// Replace with placeholder if outside buffer
+							return null;
+						} )
+						.filter(
+							( record, index ) =>
+								record !== null ||
+								( newMin + index >= visibleMin - buffer &&
+									newMin + index <= visibleMax + buffer )
+						);
+
+					return filtered;
+				}
+
+				return result.filter( ( r ) => r !== null );
+			} );
+		}
+		setIsLoadingMore( false );
+	}, [
+		shownData,
+		view.search,
+		view.filters,
+		view.perPage,
+		currentPage,
+		view.infiniteScrollEnabled,
+		allLoadedRecords.length,
+		scrollDirection,
+		visibleEntries,
+		getItemId,
+	] );
+
+	const paginationInfo = {
+		totalItems,
+		totalPages,
+		infiniteScrollHandler,
+		setVisibleEntries,
+	};
+
+	// Filter out null placeholders for display
+	const displayData = allLoadedRecords.filter(
+		( record ): record is Item => record !== null
+	);
```

Comment:
I've only looked at the story so I could be mistaken, but I'm seeing mixed results in the `n of nTotal items loaded` string.

The story uses `displayData.length` for `n` - does `displayData` represent all loaded items or just those within the visible range?

E.g., When I scroll to the bottom I'm seeing `13 of 37 items loaded.All items loaded!`, sometimes it's `10 of 37`. And when I scroll to the top, it changes to `15`. What's expected?